### DISCUSSION TITLE
Question about `colour.sd_to_XYZ` definition.

### DISCUSSION THREAD

**Discussion by xkennyz** [colour-science/colour#1247] [Q&A] [answered]
Hi all!

First time writing here. 
I just wanted to know if it's possible to make it so that the colour.sd_to_XYZ _only_ integrates over the inputted wavelengths. For example, if I have a spectral distribution ranging from 500 to 700nm, the function will extrapolate the first and last values linearly to the left and right, respectively. _Then_ it will perform the calculation.
I've tried setting the other wavelengths to zero (e.g., 495nm and below to 0, 705nm and above to 0), but the function doesn't quite give me what I expect it to. I believe it's because there's some sort of averaging that the function does to integrate (between 495 and 500, 700 and 705).
Basically, I want to somehow find a way to limit the bounds of integration to the range of my spectral distribution rather than it being 360~830 indefinitely.
I'm a little new to this stuff, so I'm sorry if I missed something basic.

Thanks!

**Comment by xkennyz** [comment]
Never mind! I modified the source code to do what I wanted.
Thanks for the great package!

**Comment by tjdcs** [comment]
Use `sd_to_XYZ(sd.values, shape=sd.shape)` and also `sd_to_XYZ(sd.values, shape=sd.shape, method="Integration")` should do what you want. However, the default behavior is is based on CIE recommendations and other standard practices. In my own work I use the above function calls because they also use a faster code path, but they are considered less accurate. 

Your example is relatively bizarre, I'm curious why you don't have any data below 500nm? In this case I don't think there is any right answer as so much of the visual system's sensitivity is below this wavelength. It's hard to say any colorimetric conversions are meaningful.

**Reply by xkennyz** [reply]
Haha sorry, I was just giving an example of random wavelengths. In hindsight, it does look suspect.
The real data I'm dealing with goes from 480 to 620nm. Not entirely sure why I didn't just say that.

As for putting in the shapes, I've found that no matter what shape I put in, the spectral distribution gets extrapolated out to 360 and 780.
When I looked into the source code, I found that what happens is the spectral distribution itself gets reshaped into the shape of cmfs, RATHER THAN reshaping cmfs to the spectral distribution, which explains why nothing I did was able to get the function to work--it was working, just not the way I expected it to.

**Reply by tjdcs** [reply]
Yes. That is the recommended procedure by the CIE. However, the code path I posted above will adhere to your preference.

**Reply by xkennyz** [reply]
Sadly, as I mentioned above, it does not replicate the result I want.

[data.txt](https://github.com/colour-science/colour/files/14563728/data.txt)
Using the data I have as an example (attached), when I run your suggestions, it comes out to be:
     XYZ with sd_to_XYZ is: [  753.45524368  2268.18722396   207.25291574]

The values I'm looking for are:
     XYZ is: [ 743.2332, 2264.1323, 206.3468 ]

Using the code I modified, I get:
     XYZ is: [  743.51760662  2264.25109939   206.3403274 ]

Which is close enough to what I wanted.

**Reply by thomasmansencal** [reply, marked as answer]
It is because `shape` is only used when integrating a n-dimensional array and not when using the `colour.continouous.AbstractContinuousSignal` sub-classes, e.g., `colour.SpectralDistribution`. The reason why is it like that is because integrating n-dimensional arrays was implemented much later and it did not occur to me that shape could/should also drive the integration of the initial code path but I think it should, it totally makes sense!

**Reply by thomasmansencal** [reply]
(Note that the docs are explicit about when `shape` is being used though)

**Reply by xkennyz** [reply]
Huh. Go figure. Thanks, that explains it!

**Reply by tjdcs** [reply]
I don't get the same results as you with the data.txt you provided, but here is a notebook to explore the difference more. As @KelSolaar said there is a switch to force the reshaping of the cmfs rather than reshaping the input data based on passing in an n-array instead of an sd here: 

https://github.com/colour-science/colour/blob/7fb9fccd020255604d059d05c287974a05f9c888/colour/colorimetry/tristimulus_values.py#L677-L715

Unless I'm still misunderstanding @xkennyz variables. I tried debugging with the example data but ended up with totally different XYZ values. 

```python
import numpy as np
from colour.colorimetry.spectrum import SpectralDistribution, SpectralShape
from colour.colorimetry.tristimulus_values import sd_to_XYZ

data = [
    0.001447728,
    0.031316602,
    0.281427964,
    1.164928271,
    2.827705839,
    4.683560365,
    6.304900832,
    7.418446126,
    7.92553908,
    7.948622573,
    7.571148614,
    6.952271141,
    6.218380217,
    5.365852514,
    4.528914745,
    3.753913463,
    3.028762779,
    2.392047624,
    1.861467159,
    1.419677822,
    1.066408985,
    0.789787672,
    0.576711539,
    0.415404445,
    0.296450351,
    0.208676513,
    0.144236724,
    0.100071995,
    0.068515426,
]

shape = SpectralShape(480, 620, 5)
sd = SpectralDistribution(data, domain=shape)

# This code path reshapes the cmfs to the input sd.shape
#
# The switch is on line 662 of tristimulus_values.py based on the type of the
# first argument

XYZ1 = sd_to_XYZ(sd.values, shape=sd.shape, method="Integration")

# This code path reshapes sd to match the default CMFs.

XYZ2 = sd_to_XYZ(sd)

print("XYZ1: " + np.array2string(XYZ1, formatter={"float_kind": lambda x: "%.4f" % x}))
print("XYZ2: " + np.array2string(XYZ2, formatter={"float_kind": lambda x: "%.4f" % x}))
```

```
> python test.py
XYZ1: [112.7542 343.0918 31.2653]
XYZ2: [103.2273 310.7786 28.3986]
```

I use this code path pretty often because it is faster during a lot of the optimization code I write, if it's not behaving as described then I'm really confused and pretty concerned about a lot of my results. Ha.

**Reply by tjdcs** [reply]
note* instead of passing in sd.values you could pass in the data variable directly but there is a bug right now where it fails to calculate a cache key correctly (need to make data an `np.ndarray`). I'll see if theres a nice place to insert a fix shortly. 

```python
XYZ1 = sd_to_XYZ(np.asarray(data), shape=sd.shape, method="Integration")
```

**Reply by xkennyz** [reply]
Sorry, I should have mentioned that I put k=6.83.

Your code does work, actually! When I ran it, this is what I got:
     XYZ is (my modified code): [  743.51760662  2264.25109939   206.3403274 ]
     XYZ1 with sd_to_XYZ is: [  744.23305155  2264.57473479   206.36653627]
     XYZ2 with sd_to_XYZ is: [  753.38642634  2268.16284154   207.26204103]

Note that for the first result, I'm still using sd_to_XYZ_integration. This is what I modified the code to be:

````
def sd_to_XYZ_integration(
    sd: ArrayLike | SpectralDistribution | MultiSpectralDistributions,
    cmfs: MultiSpectralDistributions | None = None,
    illuminant: SpectralDistribution | None = None,
    k: Real | None = None,
    shape: SpectralShape | None = None,
) -> NDArrayFloat:
    """
    Convert given spectral distribution to *CIE XYZ* tristimulus values
    using given colour matching functions and illuminant according to classical
    integration method.

    The spectral distribution can be either a
    :class:`colour.SpectralDistribution` class instance or an `ArrayLike` in
    which case the ``shape`` must be passed.

    Parameters
    ----------
    sd
        Spectral distribution, if an `ArrayLike` the wavelengths are
        expected to be in the last axis, e.g. for a spectral array with
        77 bins, ``sd`` shape could be (77, ) or (1, 77).
    cmfs
        Standard observer colour matching functions, default to the
        *CIE 1931 2 Degree Standard Observer*.
    illuminant
        Illuminant spectral distribution, default to *CIE Illuminant E*.
    k
        Normalisation constant :math:`k`. For reflecting or transmitting object
        colours, :math:`k` is chosen so that :math:`Y = 100` for objects for
        which the spectral reflectance factor :math:`R(\\lambda)` of the object
        colour or the spectral transmittance factor :math:`\\tau(\\lambda)` of
        the object is equal to unity for all wavelengths. For self-luminous
        objects and illuminants, the constants :math:`k` is usually chosen on
        the grounds of convenience. If, however, in the CIE 1931 standard
        colorimetric system, the :math:`Y` value is required to be numerically
        equal to the absolute value of a photometric quantity, the constant,
        :math:`k`, must be put equal to the numerical value of :math:`K_m`, the
        maximum spectral luminous efficacy (which is equal to
        683 :math:`lm\\cdot W^{-1}`) and :math:`\\Phi_\\lambda(\\lambda)` must
        be the spectral concentration of the radiometric quantity corresponding
        to the photometric quantity required.
    shape
        Spectral shape of the spectral distribution, ``cmfs`` and
        ``illuminant`` will be aligned to it if ``sd`` is an `ArrayLike`.

    Returns
    -------
    :class:`numpy.ndarray`
        *CIE XYZ* tristimulus values.

    Notes
    -----
    +-----------+-----------------------+---------------+
    | **Range** | **Scale - Reference** | **Scale - 1** |
    +===========+=======================+===============+
    | ``XYZ``   | [0, 100]              | [0, 1]        |
    +-----------+-----------------------+---------------+

    -   When :math:`k` is set to a value other than *None*, the computed
        *CIE XYZ* tristimulus values are assumed to be absolute and are thus
        converted from percentages by a final division by 100.
    -   The code path using the `ArrayLike` spectral distribution produces
        results different to the code path using a
        :class:`colour.SpectralDistribution` class instance: the former
        favours execution speed by aligning the colour matching functions and
        illuminant to the given spectral shape while the latter favours
        precision by aligning the spectral distribution to the colour matching
        functions.

    References
    ----------
    :cite:`Wyszecki2000bf`

    Examples
    --------
    >>> from colour import MSDS_CMFS, SDS_ILLUMINANTS, SpectralDistribution
    >>> cmfs = MSDS_CMFS["CIE 1931 2 Degree Standard Observer"]
    >>> illuminant = SDS_ILLUMINANTS["D65"]
    >>> shape = SpectralShape(400, 700, 20)
    >>> data = np.array(
    ...     [
    ...         0.0641,
    ...         0.0645,
    ...         0.0562,
    ...         0.0537,
    ...         0.0559,
    ...         0.0651,
    ...         0.0705,
    ...         0.0772,
    ...         0.0870,
    ...         0.1128,
    ...         0.1360,
    ...         0.1511,
    ...         0.1688,
    ...         0.1996,
    ...         0.2397,
    ...         0.2852,
    ...     ]
    ... )
    >>> sd = SpectralDistribution(data, shape)
    >>> sd_to_XYZ_integration(sd, cmfs, illuminant)
    ... # doctest: +ELLIPSIS
    array([ 10.8404805...,   9.6838697...,   6.2115722...])
    >>> sd_to_XYZ_integration(data, cmfs, illuminant, shape=shape)
    ... # doctest: +ELLIPSIS
    array([ 10.8993917...,   9.6986145...,   6.2540301...])

    The default CMFS are the *CIE 1931 2 Degree Standard Observer*, and the
    default illuminant is *CIE Illuminant E*:

    >>> sd_to_XYZ_integration(sd)
    ... # doctest: +ELLIPSIS
    array([ 11.7786939...,   9.9583972...,   5.7371816...])
    """

    as_percentage = k is not None

    # NOTE: The "illuminant" argument is reshaped by the
    # `handle_spectral_arguments` definition, but, in this case, it is not
    # desirable as we want to reshape it according to the final "shape" which
    # is only available after the subsequent if/else block thus we are careful
    # not unpacking over it.
    if illuminant is None:
        cmfs, illuminant = handle_spectral_arguments(
            cmfs, illuminant, illuminant_default="E"
        )
    else:
        cmfs, _illuminant = handle_spectral_arguments(
            cmfs, illuminant, illuminant_default="E"
        )

    if isinstance(sd, (SpectralDistribution, MultiSpectralDistributions)):
        # since the array begins with the first value corresponding to 360nm, i have to subtract 360 
        # from the starting wavelength of my spectral distribution! that way, it will start at the right wavelength.
        begin = int(list(sd)[0][0])-360
        end = int(list(sd)[-1][0])-360

        shape = cmfs.shape

        if sd.shape != shape:
            runtime_warning(
                f'Aligning "{sd.name}" spectral data shape to "{shape}".'
            )

            sd = (
                reshape_sd(sd, shape, copy=False)
                if isinstance(sd, SpectralDistribution)
                else reshape_msds(sd, shape, copy=False)
            )

        R = np.transpose(sd.values)
        shape_R = R.shape

        wl_c_r = R.shape[-1]
        print(wl_c_r)
    else:
        attest(
            shape is not None,
            "A spectral shape must be explicitly passed with a spectral data "
            "array!",
        )

        shape = cast(SpectralShape, shape)

        R = as_float_array(sd)
        shape_R = R.shape
        wl_c_r = R.shape[-1]
        wl_c = len(shape.wavelengths)

        attest(
            wl_c_r == wl_c,
            f"Spectral data array with {wl_c_r} wavelengths is not compatible "
            f"with spectral shape with {wl_c} wavelengths!",
        )

        if cmfs.shape != shape:
            runtime_warning(f'Aligning "{cmfs.name}" cmfs shape to "{shape}".')
            cmfs = reshape_msds(cmfs, shape, copy=False)

    if illuminant.shape != shape:
        runtime_warning(
            f'Aligning "{illuminant.name}" illuminant shape to "{shape}".'
        )
        illuminant = reshape_sd(illuminant, shape, copy=False)

    XYZ_b = cmfs.values[begin:end+1]
    S = illuminant.values[begin:end+1]
    # print(len(R))

    R = np.reshape(R, (-1, wl_c_r))[0][begin:end+1]
    # print(len(R))
    # print(R)


    d_w = cmfs.shape.interval

    with sdiv_mode():
        k = cast(
            Real, optional(k, sdiv(100, (np.sum(XYZ_b[..., 1] * S) * d_w)))
        )

    XYZ = k * np.dot(R * S, XYZ_b) * d_w

    XYZ = from_range_100(np.reshape(XYZ, [*list(shape_R[:-1]), 3]))

    if as_percentage:
        XYZ /= 100

    return XYZ
````

**Reply by thomasmansencal** [reply]
With https://github.com/colour-science/colour/pull/1250, I get `[  744.23305155  2264.57473479   206.36653627]`

```python
import colour

D = """
0.001447728
0.031316602
0.281427964
1.164928271
2.827705839
4.683560365
6.304900832
7.418446126
7.92553908
7.948622573
7.571148614
6.952271141
6.218380217
5.365852514
4.528914745
3.753913463
3.028762779
2.392047624
1.861467159
1.419677822
1.066408985
0.789787672
0.576711539
0.415404445
0.296450351
0.208676513
0.144236724
0.100071995
0.068515426
""".strip().splitlines()
D = list(map(float, D))

shape = colour.SpectralShape(480, 620, 5)
print(len(D), len(shape))

sd = colour.SpectralDistribution(D, shape)

print(colour.sd_to_XYZ(sd, k=6.83, method="Integration", shape=shape))
```
